services/galley/src/Galley/Data.hs

```diff
@@ -85,17 +108,131 @@ import qualified System.Logger.Class  as Log
+teamConversations t = map (uncurry newTeamConversation) <$>
+    retry x1 (query Cql.selectTeamConvs (params Quorum (Identity t)))
+
+teamMembers :: MonadClient m => TeamId -> m [TeamMember]
+teamMembers t = map (uncurry newTeamMember) <$>
+    retry x1 (query Cql.selectTeamMembers (params Quorum (Identity t)))
+
+teamMember :: MonadClient m => TeamId -> UserId -> m (Maybe TeamMember)
+teamMember t u = fmap (newTeamMember u . runIdentity) <$>
+    retry x1 (query1 Cql.selectTeamMember (params Quorum (t, u)))
+
+userTeams :: MonadClient m => UserId -> m [TeamId]
+userTeams u = map runIdentity <$>
+    retry x1 (query Cql.selectUserTeams (params Quorum (Identity u)))
+
+createTeam :: MonadClient m
+           => UserId
+           -> Range 1 256 Text
+           -> Range 1 256 Text
+           -> Maybe (Range 1 256 Text)
+           -> m Team
+createTeam uid (fromRange -> n) (fromRange -> i) k = do
+    tid <- Id <$> liftIO nextRandom
+    retry x5 $ write Cql.insertTeam (params Quorum (tid, uid, n, i, fromRange <$> k))
+    pure (newTeam tid uid n i & teamIconKey .~ (fromRange <$> k))
+
+deleteTeam :: MonadClient m => TeamId -> m ()
+deleteTeam tid = do
+    retry x5 $ write Cql.markTeamDeleted (params Quorum (Identity tid))
+    mm <- teamMembers tid
+    for_ mm $ removeTeamMember tid . view userId
+    cc <- teamConversations tid
+    for_ cc $ removeTeamConv tid . view conversationId
+    retry x5 $ write Cql.deleteTeam (params Quorum (Identity tid))
+
+addTeamMember :: MonadClient m => TeamId -> TeamMember -> m ()
+addTeamMember t m =
+    retry x5 $ batch $ do
+        setType BatchLogged
+        setConsistency Quorum
+        addPrepQuery Cql.insertTeamMember (t, m^.userId, m^.permissions)
+        addPrepQuery Cql.insertUserTeam   (m^.userId, t)
+
+removeTeamMember :: MonadClient m => TeamId -> UserId -> m ()
+removeTeamMember t m =
+    retry x5 $ batch $ do
+        setType BatchLogged
+        setConsistency Quorum
+        addPrepQuery Cql.deleteTeamMember (t, m)
+        addPrepQuery Cql.deleteUserTeam   (m, t)
+
+removeTeamConv :: MonadClient m => TeamId -> ConvId -> m ()
+removeTeamConv tid cid = do
+    deleteConversation cid
+    retry x5 $ batch $ do
+        setType BatchLogged
+        setConsistency Quorum
+        addPrepQuery Cql.deleteTeamConv (tid, cid)
+
+updateTeam :: MonadClient m => TeamId -> TeamUpdateData -> m ()
+updateTeam tid u = retry x5 $ batch $ do
+    setType BatchLogged
+    setConsistency Quorum
+    for_ (u^.nameUpdate) $ \n ->
+        addPrepQuery Cql.updateTeamName (n, tid)
+    for_ (u^.iconUpdate) $ \i ->
+        addPrepQuery Cql.updateTeamIcon (i, tid)
+    for_ (u^.iconKeyUpdate) $ \k ->
+        addPrepQuery Cql.updateTeamIconKey (k, tid)
+
+-- Conversations ------------------------------------------------------------
+
+isConvAlive :: MonadClient m => ConvId -> m Bool
+isConvAlive cid = do
+    result <- retry x1 (query1 Cql.isConvDeleted (params Quorum (Identity cid)))
```

Comment:
Given the read / write ratio of this flag, wouldn't it be preferable to use consistency `One` here and, if really necessary, consistency `All` for the write?